_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,782 @@
+That will give us the predicted mean, as well as the predicted variance, _per datapoint_.
+
+```{r}
+mean <- yhat %>% tfd_mean()
+sd <- yhat %>% tfd_stddev()
+```
+
+Let's visualize this. Here are the actual test data points, the predicted means, as well as confidence bands indicating the mean estimate plus/minus two standard deviations.
+
+```{r}
+ggplot(data.frame(
+  x = x,
+  y = y,
+  mean = as.numeric(mean),
+  sd = as.numeric(sd)
+),
+aes(x, y)) +
+  geom_point() +
+  geom_line(aes(x = x_test, y = mean), color = "violet", size = 1.5) +
+  geom_ribbon(aes(
+    x = x_test,
+    ymin = mean - 2 * sd,
+    ymax = mean + 2 * sd
+  ),
+  alpha = 0.2,
+  fill = "grey")
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Aleatoric uncertainty on simulated data, using relu activation in the first dense layer."}
+knitr::include_graphics("images/g_aleatoric_relu_8.png")
+```
+
+
+This looks pretty reasonable. What if we had used linear activation in the first layer? Meaning, what if the model had looked like this^[yes, we also use that other line for `scale` that was commented before; more on that in a second]:
+
+```{r}
+model <- keras_model_sequential() %>%
+  layer_dense(units = 8, activation = "linear") %>%
+  layer_dense(units = 2, activation = "linear") %>%
+  layer_distribution_lambda(function(x)
+    tfd_normal(loc = x[, 1, drop = FALSE],
+               scale = 1e-3 + 0.05 * tf$math$softplus(x[, 2, drop = FALSE])
+               )
+  )
+
+```
+
+This time, the model does not capture the "form" of the data that well, as we've disallowed any nonlinearities.
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Aleatoric uncertainty on simulated data, using linear activation in the first dense layer."}
+knitr::include_graphics("images/g_aleatoric_linear_8.png")
+```
+
+Using linear activations only, we also need to do more experimenting with the `scale = ...` line to get the result look "right". With `relu`, on the other hand, results are pretty robust to changes in how `scale` is computed. Which activation do we choose? If our goal is to adequately model variation in the data, we can just choose `relu` -- and leave assessing _uncertainty in the model_ to a different technique (the _epistemic_ uncertainty that is up next).
+
+Overall, it seems like aleatoric uncertainty is the straightforward part. We want the network to learn the variation inherent in the data, which it does. What do we gain? Instead of obtaining just point estimates, which in this example might turn out pretty bad in the two fan-like areas of the data on the left and right sides, we learn about the spread as well. We'll thus be appropriately cautious depending on what input range we're making predictions for.
```

Comment:
```suggestion
Overall, it seems like aleatoric uncertainty is the straightforward part. We want the network to learn the variation inherent in the data, which it does. What do we gain? Instead of obtaining just point estimates, which in this example might turn out pretty bad in the two fan-like areas of the data on the left and right sides, we learn about the spread as well. We'll thus be appropriately cautious, depending on what input range we're making predictions for.
```